include/nbl/ext/ImGui/ImGui.h

```diff
@@ -1,76 +1,133 @@
 {
 	public:
-		UI(core::smart_refctd_ptr<video::ILogicalDevice> device, int maxFramesInFlight, video::IGPURenderpass* renderpass, video::IGPUPipelineCache* pipelineCache, core::smart_refctd_ptr<ui::IWindow> window);
+		struct MDI
+		{
+			using COMPOSE_T = nbl::video::StreamingTransientDataBufferST<nbl::core::allocator<uint8_t>>;								//! composes memory available for the general purpose allocator to suballocate memory ranges								
+			using SUBALLOCATOR_TRAITS_T = nbl::core::address_allocator_traits<nbl::core::LinearAddressAllocatorST<uint32_t>>;			//! traits for MDI buffer suballocator - fills the data given the mdi allocator memory request
+
+			enum E_BUFFER_CONTENT : uint8_t
+			{
+				EBC_DRAW_INDIRECT_STRUCTURES,
+				EBC_ELEMENT_STRUCTURES,
+				EBC_INDEX_BUFFERS,
+				EBC_VERTEX_BUFFERS,
+
+				EBC_COUNT,
+			};
+
+			nbl::core::smart_refctd_ptr<typename COMPOSE_T> buffer;							//! streaming mdi buffer
+
+			static constexpr auto MDI_BUFFER_REQUIRED_ALLOCATE_FLAGS = nbl::core::bitflag<nbl::video::IDeviceMemoryAllocation::E_MEMORY_ALLOCATE_FLAGS>(nbl::video::IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT); //! required flags
+			static constexpr auto MDI_BUFFER_REQUIRED_USAGE_FLAGS = nbl::core::bitflag(nbl::asset::IBuffer::EUF_INDIRECT_BUFFER_BIT) | nbl::asset::IBuffer::EUF_INDEX_BUFFER_BIT | nbl::asset::IBuffer::EUF_VERTEX_BUFFER_BIT | nbl::asset::IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT; //! required flags
+		};
+
+		struct S_CREATION_PARAMETERS
+		{
+			struct S_RESOURCE_PARAMETERS
+			{
+				nbl::video::IGPUPipelineLayout* const pipelineLayout = nullptr;				//! optional, default layout used if not provided declaring required UI resources such as textures (required font atlas + optional user defined textures) & corresponding samplers
+				uint32_t count = 0x45u;														//! amount of total UI textures (and corresponding samplers)
+
+				struct S_BINDING_REQUEST_INFO												//! for a given pipeline layout we need to know what is intended for UI resources
+				{
+					uint32_t setIx,															//! descriptor set index for a resource	
+					bindingIx;																//! binding index for a given resource
+				};
+
+				const S_BINDING_REQUEST_INFO textures = { .setIx = 0u, .bindingIx = 0u },	//! optional, default texture binding request info used if not provided (set & binding index)
+				samplers = { .setIx = 0u, .bindingIx = 1u };								//! optional, default sampler binding request info used if not provided (set & binding index)		
+
+				using binding_flags_t = nbl::video::IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS;
+				static constexpr auto TEXTURES_REQUIRED_CREATE_FLAGS = nbl::core::bitflag(binding_flags_t::ECF_UPDATE_AFTER_BIND_BIT) | binding_flags_t::ECF_PARTIALLY_BOUND_BIT | binding_flags_t::ECF_UPDATE_UNUSED_WHILE_PENDING_BIT; //! required flags
+				static constexpr auto SAMPLERS_REQUIRED_CREATE_FLAGS = nbl::core::bitflag(binding_flags_t::ECF_NONE); //! required flags
```

Comment:
unless you write the sampler right away and never touch it, it needs at least UPDATE_AFTER_BIND